AWS: documentation page for AWS module (#1891)
apache/iceberg: site/docs/aws.md

<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->
 
# Iceberg AWS Integrations

Iceberg provides integration with different AWS services through the `iceberg-aws` module. 
This section describes how to use Iceberg with AWS.

## Runtime Packages

The `iceberg-aws` module is bundled with Spark and Flink engine runtimes.
However, the AWS clients are not bundled so that you can use the same version as your application.

> Please note that we use the new AWS v2 SDK instead of v1.

Reviewer: How about "You will need to provide the AWS v2 SDK because that is what Iceberg depends on." Saying "instead of v1" makes it sound like you can't use both, so I would omit that as well.

You can choose to the [AWS SDK bundle](https://mvnrepository.com/artifact/software.amazon.awssdk/bundle), 
or individual AWS client packages (Glue, S3, DynamoDB, KMS) if you would like to have a minimum dependency footprint.

> For example, to use AWS features with Spark 3 and AWS clients version 2.15.40, you can start the SQL shell with:

Reviewer: You may want to call out the part that matters here, that this is adding `software.amazon.awssdk:bundle:2.15.40` to the `--packages` argument.

```sh
spark-sql --packages org.apache.iceberg:iceberg-spark3-runtime:0.11.0,software.amazon.awssdk:bundle:2.15.40 \
    --conf spark.sql.catalog.my_catalog=org.apache.iceberg.spark.SparkCatalog \
    --conf spark.sql.catalog.my_catalog.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog \

>     --conf spark.sql.catalog.my_catalog.warehouse=s3://my-bucket/my-key-prefix \

Reviewer: Does the Glue catalog support custom database locations?

Author: good point, let me add that section

    --conf spark.sql.catalog.my_catalog.gluecatalog.lock.table=myGlueLockTable
```

## Glue Catalog

Iceberg enables the use of [AWS Glue](https://aws.amazon.com/glue) as the `Catalog` implementation.
When used, an Iceberg `Namespace` is stored as a [Glue Database](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-databases.html), 

> an Iceberg `Table` is stored as a [Glue Table](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html),

Reviewer: I would drop references to Iceberg classes here. Most readers are going to care about tables through a SQL or DataFrame interface, not through Iceberg's API. The mapping to Glue database and table names is probably pretty clear.

> an Iceberg `Snapshot` is stored as a [Glue TableVersion](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html#aws-glue-api-catalog-tables-TableVersion). 

Reviewer: Is this true? I thought that only one version of the table was kept and that it pointed to the Iceberg root metadata file. That file contains more than one snapshot.

Author: Snapshot is probably a bad word, I will just use table version then. 
But Glue also stores all historical table versions for each update.

You can start using Glue catalog by specifying the `catalog-impl` as `org.apache.iceberg.aws.glue.GlueCatalog`. 

> More details about loading the catalog can be found in individual engine pages, such as [Spark](../spark/#loading-a-custom-catalog) and [Flink](../flink/#creating-catalogs-and-using-catalogs).

Reviewer: I would definitely give a full example of using `GlueCatalog` here, in addition to linking to Spark and Flink pages.

Reviewer: Sorry, there's one above. Maybe just point the reader to it.

> ### Glue Catalog ID

Reviewer: What about adding a table of configuration options instead of sections?

Author: The reason I am doing this is because the explanation is quite long, which makes the table look very messy.

Reviewer: I think it makes sense here, but I would opt for tables in most other places. Part of the appeal is it forces you to be concise.

You can specify the Glue catalog ID through `gluecatalog.id` catalog property to point to a Glue catalog in a different AWS account.
The Glue catalog ID is your numeric AWS account ID.
If the Glue catalog is in a different region, you should configure you AWS client to point to the correct region, 
see more details in [AWS client configuration](#aws-client-configurations).

### Skip Archive

By default, Glue will store all the table versions created and user can rollback a table to any historical version if needed.
However, if you are streaming data to Iceberg, this will easily create a lot of Glue table versions.

> Therefore, it is recommended to turn off the archive feature in Glue by setting `gluecatalog.skip-archive` to false.

Reviewer: I think it would be a bit cleaner if these catalog options were shorter. this could simply be `skip-archive` in the catalog config. Similarly, the one above could be `catalog-id`.

Author: This is mostly trying to be consistent with the `s3fileio` properties by having `gluecatalog` as the config namespace. We might be able to make all of them shorter, something like `glue.id` and `s3.staging.dir`. Any thoughts on this? @danielcweeks

### DynamoDB for locking Glue tables

Glue does not have a strong guarantee over concurrent updates to a table. 
Although it throws `ConcurrentModificationException` when detecting two processes updating a table at the same time,
there is no guarantee that one update would not clobber the other update.
Therefore, DynamoDB lock is enabled by default for Glue, so that for every commit, 
`GlueCatalog` first obtains a lock using a helper DynamoDB table and then try to safely modify the Glue table.
User must specify a table name through catalog property `gluecatalog.lock.table` as the helper DynamoDB lock table to use.
It is recommend to use the same DynamoDB table for operations in the same Glue catalog,
and use a different table for a different Glue catalog in another account or region.
If the lock table with the given name does not exist in DynamoDB, a new table is created with billing mode set as [Pay-per-Request](https://aws.amazon.com/blogs/aws/amazon-dynamodb-on-demand-no-capacity-planning-and-pay-per-request-pricing).
The lock has the following additional properties:

* `gluecatalog.lock.wait-ms`:  max time to wait for lock acquisition, defaults to 3 minutes

> * `gluecatalog.lock.expire-ms`: max time a table can be locked by a process, defaults to 20 minutes

Reviewer: What about `lock-wait-ms` and `lock-expire-ms`?

## S3 FileIO

Iceberg allows users to write data to S3 through `S3FileIO`.
`GlueCatalog` by default uses this FileIO, and other catalogs can load this FileIO using the `io-impl` catalog property.

### Progressive Multipart Upload

`S3FileIO` implements a customized progressive multipart upload algorithm to upload data.
Data files are uploaded in parts in parallel as soon as each part is ready,
and each file part is deleted as soon as its upload process completes.
This provides maximized upload speed and minimized local disk usage during uploads.
Here are the configurations user can tune related to this feature:

* `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared across all output streams), defaults to the available number of processors in the system
* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, defaults to 32MB
* `s3fileio.multipart.threshold`: the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload, defaults to 1.5

> * `s3fileio.staging.dir`: the directory to hold temporary files, defaults to Java's `java.io.tmpdir` property value

Reviewer: Great info here, but I think it may be easier to maintain as a table.

Author: switched to table layout.

## S3 FileIO

Iceberg allows users to write data to S3 through `S3FileIO`.
`GlueCatalog` by default uses this FileIO, and other catalogs can load this FileIO using the `io-impl` catalog property.

### Progressive Multipart Upload

`S3FileIO` implements a customized progressive multipart upload algorithm to upload data.
Data files are uploaded by parts in parallel as soon as each part is ready,
and each file part is deleted as soon as its upload process completes.
This provides maximized upload speed and minimized local disk usage during uploads.
Here are the configurations that users can tune related to this feature:

| Property                          | Default                                            | Description                                            |
| --------------------------------- | -------------------------------------------------- | ------------------------------------------------------ |
| s3fileio.multipart.num-threads    | the available number of processors in the system   | number of threads to use for uploading parts to S3 (shared across all output streams)  |

> | s3fileio.multipart.part.size      | 32MB                                               | the size of a single part for multipart upload requests  |

Reviewer: Have we released this yet? We would normally make it `s3fileio.multipart.part-size-bytes`:
* `part` is part of "part size" and isn't really part of the hierarchy
* We prefer to have clear units

| s3fileio.multipart.threshold      | 1.5                                                | the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload  |

> | s3fileio.staging.dir              | `java.io.tmpdir` property value                    | the directory to hold temporary files  |

Reviewer: Is there anything else under staging, or should this be `staging-dir`?

Author: Looks like there are multiple discussions around the config key names, and these names for `s3fileio` are not designed by me but Daniel. Let me put up another thread for this discussion before release.

Author: let's use #2050 for the discussion

<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->
 
# Iceberg AWS Integrations

Iceberg provides integration with different AWS services through the `iceberg-aws` module. 
This section describes how to use Iceberg with AWS.

## Runtime Packages

> The first thing to note is that the `iceberg-aws` module is not bundled with any engine runtime.

Author: Yes

To use any features described in later sections, you need to include the following packages by yourself:

* the [iceberg AWS package](https://mvnrepository.com/artifact/org.apache.iceberg/iceberg-aws)

> * the [AWS SDK bundle](https://mvnrepository.com/artifact/software.amazon.awssdk/bundle), or individual AWS client packages if you would like to have a minimum dependency footprint. (please note that we use the new AWS v2 SDK instead of v1)

Author: Fixed

an Iceberg `Table` is stored as a [Glue Table](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html),
an Iceberg `Snapshot` is stored as a [Glue TableVersion](https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-tables.html#aws-glue-api-catalog-tables-TableVersion). 
You can start using Glue catalog by specifying the `catalog-impl` as `org.apache.iceberg.aws.glue.GlueCatalog`. 
More details about loading the catalog can be found in individual engine pages, such as [Spark](../spark/#loading-a-custom-catalog) and [Flink](../flink/#creating-catalogs-and-using-catalogs).

### Glue Catalog ID

It is very common for an organization to store all the tables in a single Glue catalog in a single AWS account and run data computation in many different accounts. 
In this case, you need to specify a Glue catalog ID when initializing `GlueCatalog`.
The Glue catalog ID you should use is the AWS account ID.
This is because in each AWS account, there is a single Glue catalog in each AWS region,
but the region is pre-determined by the Glue web client that is making the call.

> If you would like to access a Glue catalog in a different region, you should configure you AWS client, see more details in [AWS client configuration](#aws-client-configurations).

Author: I see, please read if the new version is easier to understand.

It is recommend to use the same DynamoDB table for operations in the same Glue catalog,
and use a different table for a different Glue catalog in another account or region.
If the lock table with the given name does not exist in DynamoDB, a new table is created with billing mode set as [Pay-per-Request](https://aws.amazon.com/blogs/aws/amazon-dynamodb-on-demand-no-capacity-planning-and-pay-per-request-pricing).
The lock has the following additional properties:

* `gluecatalog.lock.wait-ms`:  max time to wait for lock acquisition, default to 3 minutes
* `gluecatalog.lock.expire-ms`: max time a table can be locked by a process, default to 20 minutes

If your use case only consists of single-process low-frequency (e.g. hourly, daily) updates to a table,
you can also turn off this locking feature by setting `gluecatalog.lock.enabled` as false.

### Warehouse Location

By default, Glue uses `S3FileIO` and only allows a warehouse location in S3. 
To store data in a different local or cloud store, Glue catalog can switch to use `HadoopFileIO` 
or any custom FileIO using the mechanism described in the [custom FileIO](../custom-catalog/#custom-file-io-implementation) section.

> ## S3 FileIO

Reviewer 2: I was about to ask the rationalization of S3 FileIO compared to Hadoop filesystem API with S3 support in #1945, but this section covers it. Thanks!

Probably worth to also mention whether Hadoop FS API with S3 is sufficient to work with, or S3 FileIO is required to avoid consistency glitches. That would help end users to determine whether including aws module is a kind of requirement for dealing with S3 or not.

Author: I have updated strong consistency section for more details, please let me know if it is enough.

> Iceberg allows user to write data to S3 through `S3FileIO`.

Reviewer 3: ```suggestion
Iceberg allows users to write data to S3 through `S3FileIO`.
```

`GlueCatalog` by default uses this FileIO, and other catalogs can load this FileIO using the `io-impl` catalog property.

### Progressive Multipart Upload

`S3FileIO` implements a customized progressive multipart upload algorithm to upload data.
Data files are uploaded in parts in parallel as soon as each part is ready,
and each file part is deleted as soon as its upload process completes.
This provides maximized upload speed and minimized local disk usage during uploads.
Here are the configurations user can tune related to this feature:

> * `s3fileio.multipart.num-threads`: number of threads to use for uploading parts to S3 (shared pool across all output streams)

Reviewer 3: Is there a default here? Or is there no default and setting this is required?

Author: default to `Runtime.getRuntime().availableProcessors()`, let me add in the doc

> * `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, default to 32MB

Reviewer 3: ```suggestion
* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, defaults to 32MB
```
or
```suggestion
* `s3fileio.multipart.part.size`: the size of a single part for multipart upload requests, default is 32MB
```

> * `s3fileio.multipart.threshold`: the threshold expressed as a factor times the multipart size at which to switch from uploading using a single put object request to uploading using multipart upload, default to 1.5

Reviewer 3: as above `defaults to` or `default is`

> * `s3fileio.staging.dir`: the directory to hold temporary files, default to Java's `java.io.tmpdir` property value

Reviewer 3: as above to/is

* [SSE-KMS](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingKMSEncryption.html): Server-Side Encryption with Customer Master Keys (CMKs) Stored in AWS Key Management Service (SSE-KMS) is similar to SSE-S3, but with some additional benefits and charges for using this service. There are separate permissions for the use of a CMK that provides added protection against unauthorized access of your objects in Amazon S3. SSE-KMS also provides you with an audit trail that shows when your CMK was used and by whom. Additionally, you can create and manage customer managed CMKs or use AWS managed CMKs that are unique to you, your service, and your Region.
* [SSE-C](https://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html): With Server-Side Encryption with Customer-Provided Keys (SSE-C), you manage the encryption keys and Amazon S3 manages the encryption, as it writes to disks, and decryption, when you access your objects.

To enable server side encryption, use the following configuration properties:

> * `s3fileio.sse.type`: `none`, `s3`, `kms` or `custom`, default to `none`

Reviewer 3: As above for defaults.

> * `s3fileio.sse.key`: a KMS Key ID or ARN for `kms` type (default to `aws/s3`), or a custom base-64 AES256 symmetric key for `custom` type.

Author: Yes that is the default at service side.

* `s3fileio.sse.md5`: if SSE type is `custom`, this value must be set as the base-64 MD5 digest of the symmetric key to ensure integrity.

### S3 Access Control List

`S3FileIO` supports S3 access control list for detailed access control. 
User can choose the ACL level by setting the `s3fileio.acl` property.
For more details, please read [S3 ACL Documentation](https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html).

### ObjectStoreLocationProvider

S3 and many other cloud storage services [throttle requests based on object prefix](https://aws.amazon.com/premiumsupport/knowledge-center/s3-request-limit-avoid-throttling/). 
This means data stored in a traditional Hive storage layout has bad read and write throughput since data files of the same partition are placed under the same prefix.

> Iceberg by default uses the Hive storage layout, but can be switch to use a different `ObjectStoreLocationProvider`.

Reviewer 3: ```suggestion
Iceberg by default uses the Hive storage layout, but can be configured to use a different `ObjectStoreLocationProvider`.
```

In this mode, a hash string is added to the beginning of each file path, so that files are equally distributed across all prefixes in an S3 bucket.
This results in minimized throttling and maximized throughput for S3-related IO operations.

> For more details, please follow [LocationProvider Configuration](../custom-catalog/#custom-location-provider-implementation) section to enable this feature.  

Reviewer 3: ```suggestion
For more details, please refer to the [LocationProvider Configuration](../custom-catalog/#custom-location-provider-implementation) section to see how to enable this feature.  
```

### S3 Strong Consistency

In November 2020, S3 announced [strong consistency](https://aws.amazon.com/s3/consistency/) for all GET and LIST operations, and Iceberg is updated to fully leverage this feature.

> When creating a new output file using `OutputFile.create()`, strong consistency check is used and an `AlreadyExistsException` will be thrown if the file already exists in S3.

Reviewer 3: ```suggestion
When creating a new output file using `OutputFile.create()`, a strong consistency check is used and an `AlreadyExistsException` will be thrown if the file already exists in S3.
```

> ### Hadoop S3A

Reviewer 3: ```suggestion
### Hadoop S3A FileSystem
```
I think it's useful to refer to the FileSystem otherwise some people might get confused by the differences between S3, S3A, S3N etc.

Before `S3FileIO` was introduced, many Iceberg users choose to use `HadoopFileIO` to write data to S3 through the [S3A FileSystem](https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java).
As introduced in the above sections, the `S3FileIO` adopts latest AWS clients and S3 features for optimzied security and performance,

>  and is thus recommend for S3 use case over S3A.

Reviewer 3: ```suggestion
 and is thus recommend for S3 use cases rather than the S3A FileSystem.
```

`S3FileIO` is compatible with legacy URI schemes written by S3A, 
so any existing tables with `s3a://` or `s3n://` file paths are treated as equivalent `s3://` file paths.    

> If for any reason you have to use S3A, here are the instructions:

Reviewer 3: Here you are talking about storing data using `s3a://` URLs as opposed to the above discussion about the S3A FileSystem right? This should probably be reworded to be clear about that if that is the case.

Author: Thank you, I updated the section, please let me know if it is clearer.

> 1. to store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 

Reviewer 3: ```suggestion
1. To store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 
```

> 2. For `HiveCatalog`, to store metadata also using S3A, specify Hadoop config `hive.metastore.warehouse.dir` to be an S3A path.

Reviewer 3: ```suggestion
2. For `HiveCatalog`, to also store metadata using S3A, specify the Hadoop config property `hive.metastore.warehouse.dir` to be an S3A path.
```

> 3. Add [hadoop-aws](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-aws) as a runtime dependency, configure AWS settings based on [hadoop-aws documentation](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html) (make sure you check the version, s3a configuration varies a lot based on the version you use)   

Reviewer 3: ```suggestion
3. Add [hadoop-aws](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-aws) as a runtime dependency, configure AWS settings based on [hadoop-aws documentation](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html) (make sure you check the version as S3A configuration varies a lot based on the version you use) .
```

## AWS client configurations

Many organizations have customized their way of obtaining AWS credential and region information and configuring details about AWS clients for features like proxy access, retry, etc.

> Therefore, we open a configurer interface for Iceberg users to plugin any client configuration in a centralized place. 

Reviewer 3: ```suggestion
Therefore, we open a configurer interface for Iceberg users to plug in any client configuration in a centralized place. 
```

import software.amazon.awssdk.awscore.client.builder.AwsClientBuilder;
import software.amazon.awssdk.services.s3.S3ClientBuilder;

public class MyCustomClientConfigurer implements AwsClientConfigurer {

  @Override
  public <T extends AwsClientBuilder & AwsSyncClientBuilder> T configure(T clientBuilder) {
    
    // set some custom s3 configurations
    if (clientBuilder instanceof S3ClientBuilder) {
      S3ClientBuilder s3ClientBuilder = (S3ClientBuilder) clientBuilder;
      // configure something
    }

    // set the same credential provider for all clients
    clientBuilder.credentialsProvider(ContainerCredentialsProvider.builder().build());
  }
}
```

### AssumeRoleConfigurer

As a common use case scenario, we provide `AssumeRoleConfigurer` as an example configurer. It has the following catalog properties:

* `client.assume-role.arn`: role ARN to assume
* `client.assume-role.timeout-sec`: seconds for an assume-role session, after the timeout a new session is automatically fetched by a STS client.
* `client.assume-role.external-id`: optional external ID for the role to assume
* `client.assume-role.region`: a region for all clients (except the STS client) to use

> When this configurer is used, a STS client is initialized with default [credentials chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html) and [region chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html),

Reviewer 3: ```suggestion
When this configurer is used, a STS client is initialized with a default [credentials chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html) and [region chain](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html),
```

and all the other clients (Glue, DynamoDB, S3, etc.) will use the configured assume-role credential and region.

> ## Run Iceberg on AWS

Reviewer 4: Thank you @jackye1995  it looks great, this is more an question. is it also possible to run iceberg with AWS "Glue Job"? if we use `spark.jars.packages` to provide iceberg library.

Author: I have not tried that yet since I am mostly focused on Athena use case, but let me check if anyone has done that, thanks for bringing up this use case.

> [Amazon EMR](https://aws.amazon.com/emr/) is the most common platform to run Iceberg on AWS. EMR can provision clusters with [Spark](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark.html) (EMR 6 for Spark 3, EMR 5 for Spark 2),

Reviewer 5: Minor: did you have to pull down the aws sdk bundle to run sparksql in EMR? IIRC there's already some preinstalled aws libraries on EMR, if that's true we might be able to mention them here to save one step for the users.

Author: The v2 library is not bundled in spark.

[Hive](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-hive.html), [Flink](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-flink.html),
[Presto](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-presto.html) that can run Iceberg.

> [Amazon Kinesis Data Analytics](https://aws.amazon.com/about-aws/whats-new/2019/11/you-can-now-run-fully-managed-apache-flink-applications-with-apache-kafka/) also provdies a way 

Reviewer 3: ```suggestion
[Amazon Kinesis Data Analytics](https://aws.amazon.com/about-aws/whats-new/2019/11/you-can-now-run-fully-managed-apache-flink-applications-with-apache-kafka/) also provides a way 
```

> to run fully managed Apache Flink application. You can include Iceberg in your application Jar and run it in the platform.

Reviewer 3: ```suggestion
to run fully managed Apache Flink applications. You can include Iceberg in your application Jar and run it in the platform.
```

Details about this feature can be found in the [custom FileIO](../custom-catalog/#custom-file-io-implementation) section.

### Table Location

By default, the root location for a table `my_table` of namespace `my_ns` is at `my-warehouse-location/my-ns.db/my-table`.
This default root location can be changed at both namespace and table level.

To use a different path prefix for all tables under a namespace, use AWS console or any AWS Glue client SDK you like to update the `locationUri` attribute of the corresponding Glue database.
For example, you can update the `locationUri` of `my_ns` to `s3://my-ns-bucket`, 
then any newly created table will have a default root location under the new prefix.
For instance, a new table `my_table_2` will have its root location at `s3://my-ns-bucket/my_table_2`.

To use a completely different root path for a specific table, set the `location` table property to the desired root path value you want.
For example, in Spark SQL you can do:

```sql
CREATE TABLE my_catalog.my_ns.my_table (
    id bigint,
    data string,
    category string)
USING iceberg

> OPTIONS ('location'='s3://my-special-table-bucket')

Reviewer: You can also use the `LOCATION` DDL clause

Author: oh yeah I completely forgot that

To enable server side encryption, use the following configuration properties:

| Property                          | Default                                  | Description                                            |
| --------------------------------- | ---------------------------------------- | ------------------------------------------------------ |
| s3.sse.type                       | `none`                                   | `none`, `s3`, `kms` or `custom`                        |
| s3.sse.key                        | `aws/s3` for `kms` type, null otherwise  | A KMS Key ID or ARN for `kms` type, or a custom base-64 AES256 symmetric key for `custom` type.  |
| s3.sse.md5                        | null                                     | If SSE type is `custom`, this value must be set as the base-64 MD5 digest of the symmetric key to ensure integrity. |

### S3 Access Control List

`S3FileIO` supports S3 access control list (ACL) for detailed access control. 
User can choose the ACL level by setting the `s3.acl` property.
For more details, please read [S3 ACL Documentation](https://docs.aws.amazon.com/AmazonS3/latest/dev/acl-overview.html).

> ### ObjectStoreLocationProvider

Reviewer: How about using a description here rather than a class name, like "Object store file layout" or something?

Author: Yes that's better, let me also remove other class names in titles

This feature allows people to easily switch from S3A to `S3FileIO`.

If for any reason you have to use S3A, here are the instructions:

1. To store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 
2. For `HiveCatalog`, to also store metadata using S3A, specify the Hadoop config property `hive.metastore.warehouse.dir` to be an S3A path.
3. Add [hadoop-aws](https://mvnrepository.com/artifact/org.apache.hadoop/hadoop-aws) as a runtime dependency of your compute engine.
4. Configure AWS settings based on [hadoop-aws documentation](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/index.html) (make sure you check the version, S3A configuration varies a lot based on the version you use).   

## AWS client customization

Many organizations have customized their way of configuring AWS clients with their own credential provider, access proxy, retry strategy, etc.
Iceberg allows users to plug in their own implementation of `org.apache.iceberg.aws.AwsClientFactory` by setting the `client.factory` catalog property.

> ### AssumeRoleAwsClientFactory

Reviewer: Is there a better heading for this as well?

Author: Yes I just updated
